collector/benchlib/src/benchmark.rs

```diff
@@ -102,19 +107,60 @@ impl BenchmarkGroup {
 }
 
 /// Adds a single benchmark to the benchmark group.
-/// ```ignore
-/// use benchlib::define_benchmark;
 ///
-/// define_benchmark!(group, my_bench, {
-///     || do_something()
-/// });
+/// The first argument should be the `BenchmarkGroup` into which will the benchmark be stored.
+/// The second argument should be the name of the benchmark (a string literal).
+/// The third argument should be an identifier of a function that will be benchmarked.
+///
+/// If you do not provide any other arguments, the benchmarked function should not receive any
+/// parameters (it can return something though).
+///
+/// If you provide a fourth argument, it should be an expression `init` that prepares
+/// input data for the benchmarked function. The function should then receive a single parameter of
+/// the same type as `init`.
+/// ```
+/// use benchlib::{benchmark_group, define_benchmark};
+///
+/// fn calc_no_param() -> u32 { 1 + 1 }
+///
+/// fn calc_param(a: u32) -> u32 {
+///     a + 1
+/// }
+///
+/// fn main() {
+///     benchmark_group(|group| {
+///         define_benchmark!(group, my_bench, calc_no_param);
+///         define_benchmark!(group, my_bench, calc_param, 1);
+///     });
+/// }
 /// ```
 #[macro_export]
 macro_rules! define_benchmark {
-    ($group:expr, $name:ident, $fun:expr) => {
-        let func = move || $fun;
-        $group.register(stringify!($name), func);
-    };
+    // Version without init
+    ($group:expr, $name:literal, $fun:ident) => {{
+        let constructor = || {
+            move || {
+                let ret = $fun();
+
+                // Try to avoid optimizing the return value out.
+                $crate::benchmark::black_box(ret);
+            }
+        };
+        $group.register($name, constructor);
+    }};
+    // Version with init
+    ($group:expr, $name:literal, $fun:ident, $init:expr) => {{
```

Comment:
It should be possible to make this take any number of arguments with `$($init:expr),*` (or whatever the exact syntax is), like `println!` and other macros. That would avoid the need for tuples to pass multiple arguments. You might even be able to combine this macro rules with the zero-init one above.

Reply:
I wanted to do it, but my macro-fu isn't on par :laughing: I can make arbitrary number of arguments work, but I don't know how to avoid the tuples. Normally I would do something like this:

```rust
($group:expr, $name:literal, $fun:ident, $($init:expr),*) => {{
        let constructor = || {
             || fun($($init)*);
        };
```
But in this way the init expression evaluation would be measured by the harness, which I want to avoid. So instead, I'd like to do something like this:

```rust
($group:expr, $name:literal, $fun:ident, $($init:expr),*) => {{
        let constructor = || {
             let args = ($($init)*,);
             || fun(*args);
        };
```
With `*args` meaning: unroll the `args` tuple into individual arguments. But afaik something like this isn't expressible in Rust. Any ideas?

Reply:
That looks like the first example that I have sent in my previous commit I think. But in this case the init expression would be evaluated inside the measured closure, which we want to avoid.

Reply:
I see.

It makes me uncomfortable that the evaluation position is so subtle, and not visible from the call sites. E.g. this:
```rust
define_benchmark!(group, "nbody_10k", calculate_nbody, nbody::init(10000));
```
has a very different behaviour to this:
```rust
let init = nbody::init(10000);
define_benchmark!(group, "nbody_10k", calculate_nbody, init);
```
Perhaps the macro syntax could be changed so it doesn't look like a normal call, something like this:
```
define_benchmark!(group, "nbody_10k", calculate_nbody, delayed_init: init);
```
It's not great but at least makes it clearer that something non-standard is happening.

But now I'm wondering if the macro is hiding too much subtle stuff :/